tools/reviewer/cas_proxy/Proxy.java

```diff
@@ -0,0 +1,200 @@
+  // TODO(vmax): make bucket configurable
+  public static final String CAS_BUCKET_NAME = "test-bucket-cas";
+
+  static class AccessDeniedException extends Exception {}
+
+  static class Service extends CASProxyServiceGrpc.CASProxyServiceImplBase {
+    private FirestoreProtoClient client;
+    private AccessManager accessManager;
+    private HashMap<String, String> fileCache = new HashMap<>();
+
+    static class FileUploader implements StreamObserver<FileUploadRequest> {
+
+      private StreamObserver<FileUploadResponse> responseObserver;
+      private FirestoreProtoClient protoClient;
+      private ByteArrayOutputStream buffer;
+      private AccessManager accessManager;
+      private AtomicBoolean hasAccess = new AtomicBoolean(false);
+
+      public FileUploader(
+          FirestoreProtoClient protoClient,
+          AccessManager accessManager,
+          StreamObserver<FileUploadResponse> responseObserver) {
+        this.responseObserver = responseObserver;
+        this.protoClient = protoClient;
+        this.buffer = new ByteArrayOutputStream();
+        this.accessManager = accessManager;
+      }
+
+      @Override
+      public void onNext(FileUploadRequest value) {
+        if (this.accessManager.hasWriteAccess(value.getIdToken())) {
+          hasAccess.set(true);
+        }
+
+        if (!hasAccess.get()) {
+          responseObserver.onError(new AccessDeniedException());
+        }
+
+        try {
+          buffer.write(value.getContent().toByteArray());
+        } catch (IOException e) {
+          e.printStackTrace();
+          responseObserver.onError(e);
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        responseObserver.onError(t);
+      }
+
+      @Override
+      public void onCompleted() {
+        String sha256 = Hashing.sha256().hashBytes(buffer.toByteArray()).toString();
+        File tempFile;
+        try {
+          tempFile = File.createTempFile(sha256, ".tmp");
+          FileOutputStream outputStream = new FileOutputStream(tempFile);
+          outputStream.write(buffer.toByteArray());
+          outputStream.close();
+
+          protoClient.uploadTo(Proxy.CAS_BUCKET_NAME, tempFile.toPath().toString(), sha256);
+          responseObserver.onNext(FileUploadResponse.newBuilder().setSha256(sha256).build());
+        } catch (IOException e) {
+          e.printStackTrace();
+          responseObserver.onError(e);
+        }
+      }
+    }
+
+    @Override
+    public void downloadFile(
+        FileDownloadRequest request, StreamObserver<FileDownloadResponse> responseObserver) {
+
+      if (!this.accessManager.hasReadAccess(request.getIdToken(), request.getSha256())) {
+        responseObserver.onError(new AccessDeniedException());
+      }
+
+      String fileName =
+          fileCache.computeIfAbsent(
+              request.getSha256(),
+              (sha256) -> {
+                try {
+                  return client.downloadFrom(CAS_BUCKET_NAME, sha256);
```

Comment:
Where does it download to? Where is the cache located? :)

Reply:
@oferb to a temporary directory OS provides to us (you can get the path by `System.getProperty("java.io.tmpdir")`. We should probably make it configurable on next iteration.

Reply:
Let's add a TODO for it..